File: Two_Sum.cc
/*
Given an array of integers, find two numbers such that they add up to a specific target number.

The function twoSum should return indices of the two numbers such that they add up to the target, where index1 must be less than index2. Please note that your returned answers (both index1 and index2) are not zero-based.

You may assume that each input would have exactly one solution.

Input: numbers={2, 7, 11, 15}, target=9
Output: index1=1, index2=2
*/

class Solution 
{
public:
/* 暴力。超时了。思路直接
    vector<int> twoSum(vector<int> &numbers, int target) 
    {
        vector<int> result;
        
        if (numbers.size() <= 1)
            return result;
        
        for (int i = 0; i < numbers.size()-1; i++)
        {
            for (int j = i+1; j < numbers.size(); j++)
            {
                if (numbers[i] + numbers[j] == target)
                {
                    result.push_back(i+1);
                    result.push_back(j+1);
                    return result;
                }
            }
        }

        return result;
    }
*/
    vector<int> twoSum(vector<int> &numbers, int target)
    {
        
        if (numbers.size() < 2)
            return vector<int>();
        
        map<int, int> mapping; // unordered_map更好
        vector<int> result;
        
        for (int i = 0; i < numbers.size(); i++)
        {
            mapping[numbers[i]] = i; // 存储每个数对应的下标
        }
        
        for (int i = 0; i < numbers.size(); i++)
        {
            const int gap = target - numbers[i];
            
            if (mapping.find(gap) != mapping.end() && mapping[gap] > i)
            {
                result.push_back(i + 1);
                result.push_back(mapping[gap] + 1);
                break;
            }
        }
        
        return result;
    }
};

File: go/two-sum/two-sum.go
package two_sum

import "sort" // twoSum2

/*
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
*/

// map即哈希表，遍历，每个元素到map中找一下target-v，有即返回，没有则放在map中
func twoSum(nums []int, target int) []int {
	m := make(map[int]int, len(nums)) // num->index
	for i, v := range nums {
		if _, ok := m[target-v]; ok {
			return []int{m[target-v], i} // 注意先后
		}
		m[v] = i
	}
	return nil
}

// 前提是没有重复的元素，否则两个相同元素的位置会覆盖成一个
// 先排序，然后首尾向中间靠拢
func twoSum2(srcNums []int, target int) []int {
	nums := make([]int, len(srcNums))
	copy(nums, srcNums) // 为了不改变原slice，因为要排序
	m := make(map[int]int, len(nums))
	for i, v := range nums {
		m[v] = i
	}
	sort.Ints(nums)
	left, right := 0, len(nums)-1

	for left < right {
		if nums[left]+nums[right] < target {
			left++
		} else if nums[left]+nums[right] > target {
			right--
		} else {
			return []int{m[nums[left]], m[nums[right]]}
		}
	}

	return nil
}
